### PR TITLE
feat(tracing): launch_app with traces

### LINKED ISSUE #1033: Traces as a parameter to `launch_app`

### LINKED ISSUE #1037: [gql] spans connection

### PR DESCRIPTION
**Pull request by mikeldking**
resolves #1033 
resolves #1037 

Rudimentary plumbing for traces.

```graphql
{
  spans {
  edges {
    node {
      spanKind
      parentId
      context {
        traceId
        spanId
      }
    }
  }
}
}
```

## Response
```json
{
  "data": {
    "spans": {
      "edges": [
        {
          "node": {
            "spanKind": "retriever",
            "parentId": null,
            "context": {
              "traceId": "ad7c82e7-b8a2-4276-b3dd-fba51bcb60ee",
              "spanId": "8600e670-5d00-4436-92b7-9ab3e2fba90c"
            }
          }
        },
        {
          "node": {
            "spanKind": "chain",
            "parentId": "8600e670-5d00-4436-92b7-9ab3e2fba90c",
            "context": {
              "traceId": "ad7c82e7-b8a2-4276-b3dd-fba51bcb60ee",
              "spanId": "ba563475-0786-43da-b91f-32e480a06c4f"
            }
          }
        },
        {
          "node": {
            "spanKind": "llm",
            "parentId": "ba563475-0786-43da-b91f-32e480a06c4f",
            "context": {
              "traceId": "ad7c82e7-b8a2-4276-b3dd-fba51bcb60ee",
              "spanId": "ae9b1121-7e3f-4bc0-9d54-ef4974d136fd"
            }
          }
        },
        {
          "node": {
            "spanKind": "chain",
            "parentId": "ae9b1121-7e3f-4bc0-9d54-ef4974d136fd",
            "context": {
              "traceId": "ad7c82e7-b8a2-4276-b3dd-fba51bcb60ee",
              "spanId": "5f259efc-06b5-46c7-9cb1-d2647280d73f"
            }
          }
        },
        {
          "node": {
            "spanKind": "chain",
            "parentId": null,
            "context": {
              "traceId": "86d246c1-47e5-4ec3-8a60-597c8665f834",
              "spanId": "b0224c3b-b9b1-4429-895e-10614691adc6"
            }
          }
        },
        {
          "node": {
            "spanKind": "retriever",
            "parentId": "b0224c3b-b9b1-4429-895e-10614691adc6",
            "context": {
              "traceId": "86d246c1-47e5-4ec3-8a60-597c8665f834",
              "spanId": "3d693ee8-10fe-4cdb-9b81-5f54a94a52ba"
            }
          }
        },
        {
          "node": {
            "spanKind": "tool",
            "parentId": "3d693ee8-10fe-4cdb-9b81-5f54a94a52ba",
            "context": {
              "traceId": "86d246c1-47e5-4ec3-8a60-597c8665f834",
              "spanId": "35552cfa-18fb-4756-8748-84a3f77005a4"
            }
          }
        },
        {
          "node": {
            "spanKind": "llm",
            "parentId": null,
            "context": {
              "traceId": "c8c9e190-4f37-4529-87a3-86736fcef7a7",
              "spanId": "a3542370-651c-43f4-91c3-64b1888a1e12"
            }
          }
        },
        {
          "node": {
            "spanKind": "embedding",
            "parentId": "a3542370-651c-43f4-91c3-64b1888a1e12",
            "context": {
              "traceId": "c8c9e190-4f37-4529-87a3-86736fcef7a7",
              "spanId": "bcb5da2b-7b5c-48bf-a77f-720d02663400"
            }
          }
        },
        {
          "node": {
            "spanKind": "retriever",
            "parentId": "bcb5da2b-7b5c-48bf-a77f-720d02663400",
            "context": {
              "traceId": "c8c9e190-4f37-4529-87a3-86736fcef7a7",
              "spanId": "6583fbe8-8c30-4d9a-907d-66e171809811"
            }
          }
        },
        {
          "node": {
            "spanKind": "chain",
            "parentId": "6583fbe8-8c30-4d9a-907d-66e171809811",
            "context": {
              "traceId": "c8c9e190-4f37-4529-87a3-86736fcef7a7",
              "spanId": "1d353adc-1afc-489d-947f-a34b3a07bc59"
            }
          }
        },
        {
          "node": {
            "spanKind": "chain",
            "parentId": null,
            "context": {
              "traceId": "d344f66c-2cf8-4873-b8f6-30a80da4c1d8",
              "spanId": "ba45cfa7-3e41-4ae4-8e47-2ca90b57071f"
            }
          }
        },
        {
          "node": {
            "spanKind": "retriever",
            "parentId": "ba45cfa7-3e41-4ae4-8e47-2ca90b57071f",
            "context": {
              "traceId": "d344f66c-2cf8-4873-b8f6-30a80da4c1d8",
              "spanId": "c456a23b-63df-4407-91de-c2a3ded4d503"
            }
          }
        },
        {
          "node": {
            "spanKind": "tool",
            "parentId": "c456a23b-63df-4407-91de-c2a3ded4d503",
            "context": {
              "traceId": "d344f66c-2cf8-4873-b8f6-30a80da4c1d8",
              "spanId": "905c4548-7f9e-4147-a648-e99a3dd65bcc"
            }
          }
        },
        {
          "node": {
            "spanKind": "llm",
            "parentId": "905c4548-7f9e-4147-a648-e99a3dd65bcc",
            "context": {
              "traceId": "d344f66c-2cf8-4873-b8f6-30a80da4c1d8",
              "spanId": "a8785dc9-8bde-444f-ad14-35d207b9116b"
            }
          }
        },
        {
          "node": {
            "spanKind": "embedding",
            "parentId": null,
            "context": {
              "traceId": "8629cd69-cc4a-489f-ad55-d96f90ca8cba",
              "spanId": "a8674f45-a35c-4834-8315-6a2eb3f774ef"
            }
          }
        },
        {
          "node": {
            "spanKind": "llm",
            "parentId": "a8674f45-a35c-4834-8315-6a2eb3f774ef",
            "context": {
              "traceId": "8629cd69-cc4a-489f-ad55-d96f90ca8cba",
              "spanId": "7490702e-2aa1-46d1-b8d7-dc7d1135d7d6"
            }
          }
        },
        {
          "node": {
            "spanKind": "embedding",
            "parentId": "7490702e-2aa1-46d1-b8d7-dc7d1135d7d6",
            "context": {
              "traceId": "8629cd69-cc4a-489f-ad55-d96f90ca8cba",
              "spanId": "3933abeb-e8ea-45b3-854e-1a5ad93f0288"
            }
          }
        },
        {
          "node": {
            "spanKind": "llm",
            "parentId": "3933abeb-e8ea-45b3-854e-1a5ad93f0288",
            "context": {
              "traceId": "8629cd69-cc4a-489f-ad55-d96f90ca8cba",
              "spanId": "a346f6f2-524f-476b-81d6-d0edb25cdca8"
            }
          }
        },
        {
          "node": {
            "spanKind": "retriever",
            "parentId": "a346f6f2-524f-476b-81d6-d0edb25cdca8",
            "context": {
              "traceId": "8629cd69-cc4a-489f-ad55-d96f90ca8cba",
              "spanId": "3328de01-d61d-4838-a4f1-4f0d1f22dfc9"
            }
          }
        },
        {
          "node": {
            "spanKind": "llm",
            "parentId": null,
            "context": {
              "traceId": "ad6f8b88-eb9d-444a-b604-f86078d7623c",
              "spanId": "0fbfcbe2-120b-4f97-a421-162a2df28836"
            }
          }
        },
        {
          "node": {
            "spanKind": "retriever",
            "parentId": "0fbfcbe2-120b-4f97-a421-162a2df28836",
            "context": {
              "traceId": "ad6f8b88-eb9d-444a-b604-f86078d7623c",
              "spanId": "4d4a1808-4a44-49d9-b341-67c80a95d09f"
            }
          }
        },
        {
          "node": {
            "spanKind": "tool",
            "parentId": "4d4a1808-4a44-49d9-b341-67c80a95d09f",
            "context": {
              "traceId": "ad6f8b88-eb9d-444a-b604-f86078d7623c",
              "spanId": "ed8af36b-e914-4889-bb96-954e2c37cba3"
            }
          }
        },
        {
          "node": {
            "spanKind": "chain",
            "parentId": "ed8af36b-e914-4889-bb96-954e2c37cba3",
            "context": {
              "traceId": "ad6f8b88-eb9d-444a-b604-f86078d7623c",
              "spanId": "dc5ce3a3-9313-43ad-b233-9913fd47f216"
            }
          }
        },
        {
          "node": {
            "spanKind": "embedding",
            "parentId": null,
            "context": {
              "traceId": "6bde9009-0f1c-4f90-920f-cbaf2af6fe4d",
              "spanId": "f56c810f-7fb1-4290-b320-1ec4d5618781"
            }
          }
        },
        {
          "node": {
            "spanKind": "llm",
            "parentId": "f56c810f-7fb1-4290-b320-1ec4d5618781",
            "context": {
              "traceId": "6bde9009-0f1c-4f90-920f-cbaf2af6fe4d",
              "spanId": "aaf2ac86-839f-4e29-aba7-ae9953702f80"
            }
          }
        },
        {
          "node": {
            "spanKind": "chain",
            "parentId": "aaf2ac86-839f-4e29-aba7-ae9953702f80",
            "context": {
              "traceId": "6bde9009-0f1c-4f90-920f-cbaf2af6fe4d",
              "spanId": "3034b52a-da7b-423b-875c-3b1e30de34c7"
            }
          }
        },
        {
          "node": {
            "spanKind": "chain",
            "parentId": "3034b52a-da7b-423b-875c-3b1e30de34c7",
            "context": {
              "traceId": "6bde9009-0f1c-4f90-920f-cbaf2af6fe4d",
              "spanId": "28c83398-fcc9-42a3-8a85-9165534c73cb"
            }
          }
        },
        {
          "node": {
            "spanKind": "llm",
            "parentId": null,
            "context": {
              "traceId": "7976ba73-66c0-4d50-b92b-7b5f3499965a",
              "spanId": "5b71d71b-01d7-4dcd-83e2-91ebb6a89b13"
            }
          }
        },
        {
          "node": {
            "spanKind": "tool",
            "parentId": "5b71d71b-01d7-4dcd-83e2-91ebb6a89b13",
            "context": {
              "traceId": "7976ba73-66c0-4d50-b92b-7b5f3499965a",
              "spanId": "46493e80-0736-4dd3-86b7-4a54f5bcd906"
            }
          }
        },
        {
          "node": {
            "spanKind": "tool",
            "parentId": "46493e80-0736-4dd3-86b7-4a54f5bcd906",
            "context": {
              "traceId": "7976ba73-66c0-4d50-b92b-7b5f3499965a",
              "spanId": "1afe4c94-2f2d-4ed6-b726-8bd58966af2e"
            }
          }
        },
        {
          "node": {
            "spanKind": "llm",
            "parentId": "1afe4c94-2f2d-4ed6-b726-8bd58966af2e",
            "context": {
              "traceId": "7976ba73-66c0-4d50-b92b-7b5f3499965a",
              "spanId": "9c17e536-9ccf-4c96-a0e4-901f51b3dd2c"
            }
          }
        },
        {
          "node": {
            "spanKind": "llm",
            "parentId": null,
            "context": {
              "traceId": "ef0a3348-6db7-4eac-8fa0-c8f079336e9b",
              "spanId": "102b856f-e731-4416-a3e2-f7a3328371ec"
            }
          }
        },
        {
          "node": {
            "spanKind": "embedding",
            "parentId": "102b856f-e731-4416-a3e2-f7a3328371ec",
            "context": {
              "traceId": "ef0a3348-6db7-4eac-8fa0-c8f079336e9b",
              "spanId": "dae81257-9dab-468f-9d9a-a18d9f3b43cd"
            }
          }
        },
        {
          "node": {
            "spanKind": "chain",
            "parentId": "dae81257-9dab-468f-9d9a-a18d9f3b43cd",
            "context": {
              "traceId": "ef0a3348-6db7-4eac-8fa0-c8f079336e9b",
              "spanId": "6083e496-77be-4352-bc6a-b1fd6000d05e"
            }
          }
        },
        {
          "node": {
            "spanKind": "tool",
            "parentId": "6083e496-77be-4352-bc6a-b1fd6000d05e",
            "context": {
              "traceId": "ef0a3348-6db7-4eac-8fa0-c8f079336e9b",
              "spanId": "425c01fc-facd-4c26-b34c-72863d6f4a80"
            }
          }
        },
        {
          "node": {
            "spanKind": "embedding",
            "parentId": "425c01fc-facd-4c26-b34c-72863d6f4a80",
            "context": {
              "traceId": "ef0a3348-6db7-4eac-8fa0-c8f079336e9b",
              "spanId": "f7bc1062-1aea-4afd-abb9-0358ae9ff34d"
            }
          }
        },
        {
          "node": {
            "spanKind": "embedding",
            "parentId": null,
            "context": {
              "traceId": "3df5ef25-ee3f-4a3a-b70a-48cb35fc0c32",
              "spanId": "ab89ba2c-7630-41ad-b69d-2ab7b41d0543"
            }
          }
        },
        {
          "node": {
            "spanKind": "llm",
            "parentId": "ab89ba2c-7630-41ad-b69d-2ab7b41d0543",
            "context": {
              "traceId": "3df5ef25-ee3f-4a3a-b70a-48cb35fc0c32",
              "spanId": "77f77fcf-a4e1-4270-8e1d-92a86e251158"
            }
          }
        },
        {
          "node": {
            "spanKind": "retriever",
            "parentId": "77f77fcf-a4e1-4270-8e1d-92a86e251158",
            "context": {
              "traceId": "3df5ef25-ee3f-4a3a-b70a-48cb35fc0c32",
              "spanId": "d5e759bc-a4ae-4f28-a5e8-b01573442d9a"
            }
          }
        },
        {
          "node": {
            "spanKind": "llm",
            "parentId": "d5e759bc-a4ae-4f28-a5e8-b01573442d9a",
            "context": {
              "traceId": "3df5ef25-ee3f-4a3a-b70a-48cb35fc0c32",
              "spanId": "e4535dff-f52a-4e72-a1b5-b763b7bdc05c"
            }
          }
        },
        {
          "node": {
            "spanKind": "chain",
            "parentId": "e4535dff-f52a-4e72-a1b5-b763b7bdc05c",
            "context": {
              "traceId": "3df5ef25-ee3f-4a3a-b70a-48cb35fc0c32",
              "spanId": "f0648be9-b1b1-4471-b85a-c48b255b866d"
            }
          }
        },
        {
          "node": {
            "spanKind": "embedding",
            "parentId": null,
            "context": {
              "traceId": "2a2ee34e-081f-482a-8450-ddebe48e0bf5",
              "spanId": "9cdd2215-5cad-4d6c-9833-9c7ed8059e4a"
            }
          }
        },
        {
          "node": {
            "spanKind": "chain",
            "parentId": "9cdd2215-5cad-4d6c-9833-9c7ed8059e4a",
            "context": {
              "traceId": "2a2ee34e-081f-482a-8450-ddebe48e0bf5",
              "spanId": "d76f7b1e-957d-4dd9-8379-4b9204728ee3"
            }
          }
        },
        {
          "node": {
            "spanKind": "retriever",
            "parentId": "d76f7b1e-957d-4dd9-8379-4b9204728ee3",
            "context": {
              "traceId": "2a2ee34e-081f-482a-8450-ddebe48e0bf5",
              "spanId": "456f2996-4820-4096-ab77-d6dc9ee19d04"
            }
          }
        },
        {
          "node": {
            "spanKind": "retriever",
            "parentId": "456f2996-4820-4096-ab77-d6dc9ee19d04",
            "context": {
              "traceId": "2a2ee34e-081f-482a-8450-ddebe48e0bf5",
              "spanId": "ad777039-27f1-4614-9f8e-684d86ede839"
            }
          }
        },
        {
          "node": {
            "spanKind": "embedding",
            "parentId": "ad777039-27f1-4614-9f8e-684d86ede839",
            "context": {
              "traceId": "2a2ee34e-081f-482a-8450-ddebe48e0bf5",
              "spanId": "7a68a70a-b136-40c4-a78d-3d9d46dd783e"
            }
          }
        },
        {
          "node": {
            "spanKind": "llm",
            "parentId": null,
            "context": {
              "traceId": "bc25bc05-e775-4233-b4f8-cd7a86bb6d0b",
              "spanId": "c38a84fa-bb3a-471b-b972-d0e32c1ca071"
            }
          }
        },
        {
          "node": {
            "spanKind": "embedding",
            "parentId": "c38a84fa-bb3a-471b-b972-d0e32c1ca071",
            "context": {
              "traceId": "bc25bc05-e775-4233-b4f8-cd7a86bb6d0b",
              "spanId": "18a8b37c-b139-49b8-a209-8dccaea2dbed"
            }
          }
        },
        {
          "node": {
            "spanKind": "chain",
            "parentId": "18a8b37c-b139-49b8-a209-8dccaea2dbed",
            "context": {
              "traceId": "bc25bc05-e775-4233-b4f8-cd7a86bb6d0b",
              "spanId": "fe5b3251-428a-4926-9a82-e20bc8ac9da0"
            }
          }
        }
      ]
    }
  }
}
```

### CODE REVIEW

File: src/phoenix/server/api/schema.py [hunk 1]
```diff
@@ -178,6 +180,51 @@ def hdbscan_clustering(
...
+                args=ConnectionArgs(
+                    first=first,
+                    after=after if isinstance(after, Cursor) else None,
+                    last=last,
+                    before=before if isinstance(before, Cursor) else None,
+                ),
+            )
+
+        # Convert dataframe rows to Span objects
+        spans = [
```
Comment (RogerHYang) [inline src/phoenix/server/api/schema.py:204]: nit
```suggestion
        # Convert dataframe rows to Span objects
        spans = [] if info.context.traces is None else [
```

File: src/phoenix/trace/trace_dataset.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    """
+
+    dataframe: pd.DataFrame
+
+    def __init__(self, dataframe: DataFrame):
+        # Validate the the dataframe has required fields
+        columns = dataframe.columns.values
+        missing_columns = [column for column in REQUIRED_COLUMNS if column not in columns]
+        if missing_columns:
+            raise ValueError(f"The dataframe is missing some required columns: {missing_columns}")
```
Comment (RogerHYang) [inline src/phoenix/trace/trace_dataset.py:36]: nit
```suggestion
        if missing_columns := set(REQUIRED_COLUMNS) - set(df.columns):
            raise ValueError(f"The dataframe is missing some required columns: {', '.join(missing_columns)}")
```

**Review by RogerHYang — APPROVED** [review]

File: src/phoenix/trace/trace_dataset.py (hunk 2 above)
Comment (mikeldking) [reply src/phoenix/trace/trace_dataset.py]: good ol' walrus operator

File: src/phoenix/server/api/schema.py (hunk 1 above)
Comment (mikeldking) [reply src/phoenix/server/api/schema.py]: 👍

**Pull request merged**